doc/foreign-code.rst

```diff
@@ -0,0 +1,150 @@
+:ref:`include_subdirs`.
+
+The header files must have the ``.h`` extension.
+
+Installing header files
+-----------------------
+
+It is sometimes desirable to install header files with the
+library. For that you have two choices: install them explicitely with
+an :ref:`install` stanza or use the ``install_c_headers`` field of the
+:ref:`library` stanza. This field takes a list of header files names
+without the ``.h`` extension. When a library install header files,
+these are made visible to users of the library via the include search
+path.
+
+.. _foreign-sandboxing:
+
+Foreign build sandboxing
+========================
+
+When the build of a C library is too complicated to express in the
+Dune language, it is possible to simply *sandbox* an foreign
+build. Note that this method can be used to build other things, not
+just C libraries.
+
+To do that, follow the following procedure:
+- put all the foreign code in a sub-directory
+- tell Dune not to interpret configuration files in this directory via
+  an :ref:`ignored_subdirs <dune-ignored_subdirs>` stanza
+- write a custom rule that:
+  + depend on this directory recursively via :ref:`source_tree`
+  + invoke the external build system
+  + copy the C archive files (``.a``, ``.so``, ...) in main library
+    directory with a specific names (see bellow)
+- *attach* the C archive files to an OCaml library via the
+  :ref:`self_build_stubs_archive` field
+
+For instance, let's assume that you want to build a C library
+``libfoo`` using ``libfoo``'s own build system and attach it to an
+OCaml library called ``foo``.
+
+The first step is to put the sources of ``libfoo`` in your project,
+for instance in ``src/libfoo``. Then tell dune to consider
+``src/libfoo`` as raw data by writing the following in ``src/dune``:
+
+.. code:: scheme
+
+          (ignored_subdirs (libfoo))
+
+The next step is to setup the rule to build ``libfoo``. For this,
+writing the following code ``src/dune``:
+
+.. code:: scheme
+
+          (rule
+           (deps (source_tree libfoo))
+           (targets libfoo_stubs.a dllfoo_stubs.so)
+           (action (progn
+                    (chdir libfoo (run make)))
+                    (copy libfoo/libfoo.a libfoo_stubs.a)
+                    (copy libfoo/libfoo.so dllfoo_stubs.so)))
+
+Note that the rule copies the files to ``libfoo_stubs.a`` and
+``dllfoo_stubs.so``. It is important that the files produced are
+named ``lib<ocaml-lib-name>_stubs.a`` and
+``dll<ocaml-lib-name>_stubs.so``.
+
+The last step is to attach these archives to an OCaml library as
+follow:
+
+.. code:: scheme
+
+          (library
+           (name foo)
+           (self_build_stubs_archive foo))
+
+Then, whenever you use the ``foo`` library, you will also be able to
+use C functions from ``libfoo``. Note that the argument of
+``self_build_stubs_archive`` has to be the name of the library. If you
+put anything else it will be ignored, this is a bug in Dune.
```

Comment:
I'd say "This is a known bug in dune". Maybe we can open an issue and link it there?

Reply:
indeed

Reply:
done